core/arch/riscv64/mangle.c

```diff
@@ -619,22 +619,29 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
     instr_get_rel_data_or_instr_target(instr, &tgt);
     ASSERT(opnd_is_reg(dst));
     ASSERT(opnd_is_rel_addr(instr_get_src(instr, 0)));
+    bool uses_reg_stolen = instr_uses_reg(instr, dr_reg_stolen),
+         uses_reg_tp = instr_uses_reg(instr, DR_REG_TP);
 
-    ASSERT_NOT_IMPLEMENTED(!instr_uses_reg(instr, DR_REG_TP));
-
-    if (instr_uses_reg(instr, dr_reg_stolen)) {
+    if (uses_reg_stolen || uses_reg_tp) {
         dst = opnd_create_reg(DR_REG_A0);
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
     }
 
     insert_mov_immed_ptrsz(dcontext, (ptr_int_t)tgt, dst, ilist, next_instr, NULL, NULL);
 
-    if (instr_uses_reg(instr, dr_reg_stolen)) {
+    if (uses_reg_stolen) {
         PRE(ilist, next_instr,
             instr_create_save_to_tls(dcontext, DR_REG_A0, TLS_REG_STOLEN_SLOT));
         PRE(ilist, next_instr,
             instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
+    } else if (uses_reg_tp) {
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(
+                dcontext, DR_REG_A0,
+                os_tls_offset(os_get_app_tls_base_offset(TLS_REG_LIB))));
+        PRE(ilist, next_instr,
+            instr_create_restore_from_tls(dcontext, DR_REG_A0, TLS_REG0_SLOT));
```

Comment:
Is there a test we can add these cases onto, to exercise these code paths?

Reply:
I did a quick dig through and found there is not existing test for code mangling on riscv64, so I prefer to add a test (I guess it should be `suite/tests/common/allasm_riscv64_mangle`?) for all corner cases in `riscv64/mangle.c` in a separate PR.

For now, the code sequence mentioned in the description is a simple test for this.